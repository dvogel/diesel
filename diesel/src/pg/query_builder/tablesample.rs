use crate::{
    expression::{Expression, ValidGrouping},
    pg::Pg,
    query_builder::{AsQuery, AstPass, FromClause, QueryFragment, QueryId, SelectStatement},
    query_source::QuerySource,
    result::QueryResult,
    JoinTo, SelectableExpression, Table,
};

/// Indicates the sampling method for a `TABLESAMPLE method(n)` clause. The provided percentage
/// should be an integer between 0 and 100.
#[derive(Debug, Clone, Copy)]
pub enum TablesampleMethod {
    /// Use the BERNOULLI sampline method. This is row-based, slower but more accurate.
    Bernoulli(u8),

    /// Use the SYSTEM sampling method. This is page-based, faster but less accurate.
    System(u8),
}

/// Indicates the random number seed for a `TABLESAMPLE ... REPEATABLE(f)` clause.
#[derive(Debug, Clone, Copy)]
pub enum TablesampleSeed {
    /// Have PostgreSQL generate an implied rando number generator seed.
    Auto,

    /// Provide your own random number generator seed.
    Repeatable(f64),
}

/// Represents a query with a `TABLESAMPLE` clause.
#[doc(hidden)]
#[derive(Debug, Clone, Copy)]
pub struct Tablesample<S> {
    pub source: S,
    pub method: TablesampleMethod,
    pub seed: TablesampleSeed,
}

impl<S> QueryId for Tablesample<S>
where
    Self: 'static,
    S: QueryId,
{
    type QueryId = Self;
    const HAS_STATIC_QUERY_ID: bool = <S as QueryId>::HAS_STATIC_QUERY_ID;
}

impl<S> QuerySource for Tablesample<S>
where
    S: Table + Clone,
    <S as QuerySource>::DefaultSelection: ValidGrouping<()> + SelectableExpression<Tablesample<S>>,
{
    type FromClause = Self;
    type DefaultSelection = <S as QuerySource>::DefaultSelection;

    fn from_clause(&self) -> Self::FromClause {
        self.clone()
    }

    fn default_selection(&self) -> Self::DefaultSelection {
        self.source.default_selection()
    }
}

impl<S> QueryFragment<Pg> for Tablesample<S>
where
    S: QueryFragment<Pg>,
{
    fn walk_ast<'b>(&'b self, mut out: AstPass<'_, 'b, Pg>) -> QueryResult<()> {
        self.source.walk_ast(out.reborrow())?;
        let method_text = match self.method {
            TablesampleMethod::Bernoulli(p) => format!("BERNOULLI({})", p),
            TablesampleMethod::System(p) => format!("SYSTEM({})", p),
        };
        out.push_sql(format!(" TABLESAMPLE {}", method_text).as_str());
        match self.seed {
            TablesampleSeed::Auto => { /* no-op, this is the default */ }
            TablesampleSeed::Repeatable(f) => {
                out.push_sql(format!(" REPEATABLE({})", f).as_str());
            }
        }
        Ok(())
    }
}

impl<S> AsQuery for Tablesample<S>
where
    S: Table + Clone,
    <S as QuerySource>::DefaultSelection: ValidGrouping<()> + SelectableExpression<Tablesample<S>>,
{
    type SqlType = <<Self as QuerySource>::DefaultSelection as Expression>::SqlType;
    type Query = SelectStatement<FromClause<Self>>;
    fn as_query(self) -> Self::Query {
        SelectStatement::simple(self)
    }
}

impl<S, T> JoinTo<T> for Tablesample<S>
where
    S: JoinTo<T>,
    T: Table,
    S: Table,
{
    type FromClause = <S as JoinTo<T>>::FromClause;
    type OnClause = <S as JoinTo<T>>::OnClause;

    fn join_target(rhs: T) -> (Self::FromClause, Self::OnClause) {
        <S as JoinTo<T>>::join_target(rhs)
    }
}

impl<S> Table for Tablesample<S>
where
    S: Table + Clone + AsQuery,

    <S as Table>::PrimaryKey: SelectableExpression<Tablesample<S>>,
    <S as Table>::AllColumns: SelectableExpression<Tablesample<S>>,
    <S as QuerySource>::DefaultSelection: ValidGrouping<()> + SelectableExpression<Tablesample<S>>,
{
    type PrimaryKey = <S as Table>::PrimaryKey;
    type AllColumns = <S as Table>::AllColumns;

    fn primary_key(&self) -> Self::PrimaryKey {
        self.source.primary_key()
    }

    fn all_columns() -> Self::AllColumns {
        S::all_columns()
    }
}
