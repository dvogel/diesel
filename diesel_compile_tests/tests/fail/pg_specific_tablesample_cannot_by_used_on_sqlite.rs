extern crate diesel;

use diesel::dsl::*;
use diesel::query_builder::{TablesampleMethod, TablesampleSeed};
use diesel::sql_types::*;
use diesel::upsert::on_constraint;
use diesel::*;

table! {
    users {
        id -> Integer,
        name -> VarChar,
    }
}

fn main() {
    use self::users::dsl::*;
    let mut connection = SqliteConnection::establish(":memory:").unwrap();

    let random_user_ids = users
        .tablesample(
            TablesampleMethod::System(10),
            TablesampleSeed::Repeatable(42),
        )
        .load::<i32>(&mut connection);
}
