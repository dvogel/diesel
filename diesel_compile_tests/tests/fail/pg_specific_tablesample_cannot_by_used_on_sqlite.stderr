warning: unused import: `diesel::sql_types::*`
 --> tests/fail/pg_specific_tablesample_cannot_by_used_on_sqlite.rs:5:5
  |
5 | use diesel::sql_types::*;
  |     ^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused import: `diesel::upsert::on_constraint`
 --> tests/fail/pg_specific_tablesample_cannot_by_used_on_sqlite.rs:6:5
  |
6 | use diesel::upsert::on_constraint;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
  --> tests/fail/pg_specific_tablesample_cannot_by_used_on_sqlite.rs:23:41
   |
23 |             TablesampleSeed::Repeatable(42),
   |             --------------------------- ^^
   |             |                           |
   |             |                           expected `f64`, found integer
   |             |                           help: use a float literal: `42.0`
   |             arguments to this enum variant are incorrect
   |
note: tuple variant defined here
  --> $DIESEL/src/pg/query_builder/tablesample.rs
   |
   |     Repeatable(f64),
   |     ^^^^^^^^^^

error[E0277]: the trait bound `(diesel::sql_types::Integer, diesel::sql_types::Text): load_dsl::private::CompatibleType<i32, _>` is not satisfied
  --> tests/fail/pg_specific_tablesample_cannot_by_used_on_sqlite.rs:25:22
   |
25 |         .load::<i32>(&mut connection);
   |          ----        ^^^^^^^^^^^^^^^ the trait `load_dsl::private::CompatibleType<i32, _>` is not implemented for `(diesel::sql_types::Integer, diesel::sql_types::Text)`
   |          |
   |          required by a bound introduced by this call
   |
   = help: the following other types implement trait `load_dsl::private::CompatibleType<U, DB>`:
             (ST0,)
             (ST0, ST1)
             (ST0, ST1, ST2)
             (ST0, ST1, ST2, ST3)
             (ST0, ST1, ST2, ST3, ST4)
             (ST0, ST1, ST2, ST3, ST4, ST5)
             (ST0, ST1, ST2, ST3, ST4, ST5, ST6)
             (ST0, ST1, ST2, ST3, ST4, ST5, ST6, ST7)
           and $N others
   = note: required for `Tablesample<users::table>` to implement `LoadQuery<'_, _, i32>`
note: required by a bound in `diesel::RunQueryDsl::load`
  --> $DIESEL/src/query_dsl/mod.rs
   |
   |     fn load<'query, U>(self, conn: &mut Conn) -> QueryResult<Vec<U>>
   |        ---- required by a bound in this associated function
   |     where
   |         Self: LoadQuery<'query, Conn, U>,
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `RunQueryDsl::load`

error[E0277]: the trait bound `Tablesample<users::table>: QueryFragment<Sqlite>` is not satisfied
  --> tests/fail/pg_specific_tablesample_cannot_by_used_on_sqlite.rs:25:22
   |
25 |         .load::<i32>(&mut connection);
   |          ----        ^^^^^^^^^^^^^^^ the trait `QueryFragment<Sqlite>` is not implemented for `Tablesample<users::table>`
   |          |
   |          required by a bound introduced by this call
   |
   = help: the trait `QueryFragment<Pg>` is implemented for `Tablesample<S>`
   = note: required for `FromClause<Tablesample<users::table>>` to implement `QueryFragment<Sqlite>`
   = note: 2 redundant requirements hidden
   = note: required for `SelectStatement<FromClause<Tablesample<users::table>>>` to implement `QueryFragment<Sqlite>`
   = note: required for `Tablesample<users::table>` to implement `LoadQuery<'_, diesel::SqliteConnection, i32>`
note: required by a bound in `diesel::RunQueryDsl::load`
  --> $DIESEL/src/query_dsl/mod.rs
   |
   |     fn load<'query, U>(self, conn: &mut Conn) -> QueryResult<Vec<U>>
   |        ---- required by a bound in this associated function
   |     where
   |         Self: LoadQuery<'query, Conn, U>,
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `RunQueryDsl::load`
